payjoin/src/psbt/merge.rs

```diff
@@ -0,0 +1,302 @@
+//! Utilities for merging unique v0 PSBTs
+
+use std::fmt;
+
+use bitcoin::absolute::LockTime;
+use bitcoin::transaction::Version;
+use bitcoin::{OutPoint, Psbt, Transaction};
+
+const SUPPORTED_TX_VERSIONS: [i32; 1] = [2];
+
+/// Error type for merging two unique unsigned PSBTs
+#[derive(Debug)]
+pub(crate) enum MergePsbtError {
+    /// Input from other PSBT already exists in this PSBT
+    InputAlreadyExists(OutPoint),
+    /// Input is already signed
+    InputIsSigned(OutPoint),
+    /// Tx has unsupported version
+    UnsupportedTxVersion(usize, i32),
+    /// Tx has BIP68 locktime
+    LockTime(usize),
+    /// PSBT Error
+    PsbtError(bitcoin::psbt::Error),
+}
+
+impl PartialEq for MergePsbtError {
+    fn eq(&self, other: &Self) -> bool { self.to_string() == other.to_string() }
+}
+
+impl fmt::Display for MergePsbtError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::InputAlreadyExists(outpoint) =>
+                write!(f, "input already exists with outpoint: {}", outpoint),
+            Self::InputIsSigned(outpoint) =>
+                write!(f, "input is already signed with outpoint: {}", outpoint),
+            Self::UnsupportedTxVersion(index, version) =>
+                write!(f, "tx at index {} has unsupported version: {}", index, version),
+            Self::LockTime(index) => write!(f, "tx at index {} has locktime", index),
+            Self::PsbtError(e) => write!(f, "psbt error: {}", e),
+        }
+    }
+}
+
+impl std::error::Error for MergePsbtError {}
+
+#[allow(dead_code)]
+pub(crate) trait MergePsbtExt: Sized {
```

Comment:
This could use a more descriptive name now that the only method here is `dangerous_clear_signatures`. Alternatively we could add this to the main psbt ext trait